clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -731,10 +729,29 @@ private boolean targetAssignmentReconciled() {
         return currentAssignment.equals(currentTargetAssignment);
     }
 
+    /**
+     * @return True if the member should not send heartbeats, which would be one of the following
+     * cases:
+     * <ul>
+     * <li>Member is not subscribed to any topics</li>
+     * <li>Member has received a fatal error in a previous heartbeat response</li>
+     * <li>Member is stale, meaning that it has left the group due to expired poll timer</li>
+     * </ul>
+     */
     @Override
     public boolean shouldSkipHeartbeat() {
         MemberState state = state();
-        return state == MemberState.UNSUBSCRIBED || state == MemberState.FATAL;
+        return state == MemberState.UNSUBSCRIBED || state == MemberState.FATAL || state == MemberState.STALE;
+    }
+
+    /**
+     * @return True if the member is preparing to leave the group (waiting for callbacks), or
+     * leaving (sending last heartbeat). This is used to skip proactively leaving the group when
+     * the consumer poll timer expires.
+     */
+    public boolean isLeavingGroup() {
+        MemberState state = state();
+        return state == MemberState.PREPARE_LEAVING || state == MemberState.LEAVING;
     }
 
     /**
```

Comment:
I think that technically, it doesn't take the *user* to poll in order to rejoin. The code polls more frequently is the user's poll timeout is long enough.

Reply:
You're right, fixed it here and on the state description, mentioning the application poll event, which is really what it takes.